### PR TITLE
[16.0][ADD] shopinvader_api_sale_loyalty: FastAPI services to apply coupons

### PR DESCRIPTION
**Pull request by marielejeune**
FastAPI services for applying coupons to cart.

**Depends on**
* https://github.com/OCA/sale-promotion/pull/177
* https://github.com/shopinvader/odoo-shopinvader/pull/1495

**TODO**

- [x] `/apply_coupon` service must be on `cart` router, not `loyalty` router
- [x] Add a service `/apply_reward` to apply a claimable reward (automatic program, without code)
- [x] When updating cart items, we must try to apply automatic rewards if possible (a unique program, no reward choice, no product choice). The `CartResponse` must return all claimable rewards, so that before validating the cart, if some rewards are still claimable, they can be applied (by calling the `/apply_reward` service, chosing the right reward / free product).
- [x] The `shopinvader_cart_user_group` must have the `shopinvader_loyalty_user_group` as it uses all loyalties concepts.
- [x] Keep the distinction between `reward_amount` and `reward_amount_tax_included`.

### CODE REVIEW

File: shopinvader_api_sale_loyalty/schemas/cart.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+from ..schemas import LoyaltyCard, LoyaltyProgram
+
+
+class CartResponse(BaseCartResponse, extends=True):
+    promo_codes: list[str] = []
+    # TODO: in v1, we had the applied coupons in the model.
+    # Not sure it is useful as we have the coupon codes in promo_codes...
+    reward_amount: float = 0
+    programs: list[LoyaltyProgram] = []
+    generated_coupons: list[LoyaltyCard] = []
```
Comment (marielejeune) [inline shopinvader_api_sale_loyalty/schemas/cart.py]: 1. Should we add programs related info directly to the cart response?
As it is done for now, the problem is that as we extend the `CartResponse` Pydantic model, we should give read access to `shopinvader_cart_user_group` to all loyalty models.

2. I've added fields that seem useful. I don't think adding applied coupons is useful (as we have the related promo codes and the details on the sale order lines).

File: shopinvader_api_sale_loyalty/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    # TODO: needed?
+    reward_amount_tax_incl = fields.Float(compute="_compute_reward_total_tax_incl")
```
Comment (marielejeune) [inline shopinvader_api_sale_loyalty/models/sale_order.py]: In Odoo 14 the cart response returned the reward amount (tax excluded) + the reward amount tax included. In version 16 the reward_amount computation changed, it sometimes contains the amount tax included, sometimes tax excluded (depending on the type of tax on the related products).
I think returning the reward_amount (default Odoo field) is enough.

File: shopinvader_api_sale_loyalty/models/shopinvader_api_cart_service_helper.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        transactions: list[CartTransaction],
+    ):
+        # TODO: will fail when several rewards are available
+        # TODO: was a skip_coupon_recompute in params...
+        cart = super()._sync_cart(partner, cart, uuid, transactions)
+        data = LoyaltyRewardInput()
+        self.env["shopinvader_api_loyalty.service.helper"]._recompute_coupon_lines(
+            cart, data
+        )
+        return cart
```
Comment (marielejeune) [inline shopinvader_api_sale_loyalty/models/shopinvader_api_cart_service_helper.py]: In restAPI we recomputed coupon lines (to see if automatic rewards could be applied) each time an item was added to/deleted from the cart, except if a special context key was provided to skip the re-computation.
I'm not sure we should still do it here:
* We should extend Pydantic model `CardSyncInput` to take the `skip_recompute_coupons` context key.
* This re-computation will fail if an automatic loyalty program lets the user chose between several rewards. 

There exists another service `recompute_coupon_lines` that can be applied on a cart. If the convention is to call this service before validating the cart, we could improve the module to:
* For a given cart, get all automatic (i.e. without code) claimable rewards
* Specify, when recomputing coupon lines, which reward should be applied.

File: shopinvader_api_sale_loyalty/routers/loyalty.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        return reward_products[0].id
+
+    @api.model
+    def _apply_coupon(self, cart: "SaleOrder", data: LoyaltyCardInput):
+        """Apply a coupon or promotion code.
+
+        It can raise UserError if coupon is not applicable, or
+        if the coupon let the choice between several rewards, and the
+        selected reward is not specified.
+        """
```
Comment (marielejeune) [inline shopinvader_api_sale_loyalty/routers/loyalty.py]: In v16 a given coupon code can let the customer chose between several rewards.
If calling the `apply_coupon` service on such a loyalty program, without specifying which reward to apply, the request cannot succeed.
The methods in `shopinvader_api_loyalty.service.helper` raise a `UserError` in such cases. Should the service catch these errors to return a cleaner error with a specific error code?

**Comment by marielejeune** [comment]
This module will maybe need some schemas adaptations when https://github.com/shopinvader/odoo-shopinvader/pull/1421 will be finished but it's ready for first reviews.

File: shopinvader_api_sale_loyalty/routers/cart.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+from ..schemas import CartResponse, LoyaltyCardInput, LoyaltyRewardInput
+
+
+@cart_router.post("/apply_coupon")
+@cart_router.post("/apply_coupon/{cart_uuid}")
+def apply_coupon(
+    data: LoyaltyCardInput,
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/cart.py]:
```suggestion
    partner: Annotated[ResPartner, Depends(authenticated_partner)],
```

File: shopinvader_api_sale_loyalty/routers/cart.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        env["shopinvader_api_cart.service.helper"]._apply_coupon(cart, data)
+    return CartResponse.from_cart(cart) if cart else None
+
+
+@cart_router.post("/apply_reward")
+@cart_router.post("/apply_reward/{cart_uuid}")
+def apply_reward(
+    data: LoyaltyRewardInput,
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/cart.py]:
```suggestion
    partner: Annotated[ResPartner, Depends(authenticated_partner)],
```

File: shopinvader_api_sale_loyalty/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/models/sale_order.py]: @marielejeune IMO this code should got into a new addon in the `https://github.com/OCA/sale-promotion` repos.  A module named `sale_loyalty_order_info`?

File: shopinvader_api_sale_loyalty/routers/loyalty.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+from odoo import api, models
+
+from odoo.addons.fastapi.dependencies import authenticated_partner_env
+
+from ..schemas import LoyaltyRewardResponse
+
+loyalty_router = APIRouter(tags=["loyalties"])
+
+
+@loyalty_router.get("/rewards/{code}")
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/loyalty.py]: @marielejeune :thinking: `loyalties` vs `rewards` :thinking: routes prefixed with `rewards`in a router named `loyality` with a tag `loyalties`

File: shopinvader_api_sale_loyalty/routers/loyalty.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+def get_rewards(
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    code: str,
+) -> list[LoyaltyRewardResponse]:
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/loyalty.py:21]: Don't forget to put a docstring explaining what's the functionality provided by this endpoint. In which context we can expect a response etc.... The docstring is the base of the documentation provided into the swagger api for the frontend developer.

File: shopinvader_api_sale_loyalty/routers/cart.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+
+@cart_router.post("/apply_coupon")
+@cart_router.post("/apply_coupon/{cart_uuid}")
+def apply_coupon(
+    data: LoyaltyCardInput,
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
+    cart_uuid: str | None = None,
+) -> CartResponse | None:
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/cart.py]: A docstring to explain the functionality provided etc..

File: shopinvader_api_sale_loyalty/routers/cart.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+
+@cart_router.post("/apply_reward")
+@cart_router.post("/apply_reward/{cart_uuid}")
+def apply_reward(
+    data: LoyaltyRewardInput,
+    env: Annotated[api.Environment, Depends(authenticated_partner_env)],
+    partner: Annotated["ResPartner", Depends(authenticated_partner)],
+    cart_uuid: str | None = None,
+) -> CartResponse | None:
```
Comment (lmignon) [inline shopinvader_api_sale_loyalty/routers/cart.py]: A docstring to explain the functionality provided etc..

**Review by lmignon — CHANGES_REQUESTED** [review]
@marielejeune The code is clear and easy to read! A few comments to make it even better.

File: shopinvader_api_sale_loyalty/routers/loyalty.py (hunk 8 above)
Comment (marielejeune) [reply shopinvader_api_sale_loyalty/routers/loyalty.py]: This question remains open, if someone has an opinion about it.
My argument was to keep "loyalty" for the router name, as it is the base keyword for the Odoo concept.
Then the route is more specific, indicating we get all (loyalty) rewards for a given code.

File: shopinvader_api_sale_loyalty/models/sale_order.py (hunk 7 above)
Comment (marielejeune) [reply shopinvader_api_sale_loyalty/models/sale_order.py]: Done: https://github.com/OCA/sale-promotion/pull/177

**Comment by simahawk** [comment]
@marielejeune ciao, any plan to finish this?
BTW I opened a PR against your branch to fix perms, see https://github.com/acsone/odoo-shopinvader/pull/8
Could you have a look at it?

**Comment by marielejeune** [comment]
Hi @simahawk , this is finished and waiting for reviews :-) Feel free also to have a look at the dependent PR in OCA/sale-workflow

**Comment by marielejeune** [comment]
**CHANGES** (following the logic in https://github.com/shopinvader/odoo-shopinvader/pull/1495)
* Routes renaming (preserve old routes, with `deprecated=True`)
* `cart_uuid` is now of type `UUID`, not `str` anymore
* Add some doc
* Fix `_sync_cart` method: don't apply rewards on a non existing cart

**Comment by marielejeune** [comment]
> @marielejeune ciao, any plan to finish this? BTW I opened a PR against your branch to fix perms, see [acsone#8](https://github.com/acsone/odoo-shopinvader/pull/8) Could you have a look at it?

I've cherry-picked your commit

File: shopinvader_api_sale_loyalty/routers/cart.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+from odoo.addons.shopinvader_api_cart.routers import cart_router
+from odoo.addons.shopinvader_api_cart.schemas import CartTransaction
+
+from ..schemas import LoyaltyCardInput, LoyaltyRewardInput, Sale
+
+
+@cart_router.post("/apply_coupon", deprecated=True)
+@cart_router.post("/apply_coupon/{cart_uuid}", deprecated=True)
+@cart_router.post("/coupon")
+@cart_router.post("/{cart_uuid}/coupon")
```
Comment (sbidoul) [inline shopinvader_api_sale_loyalty/routers/cart.py]: @marielejeune  can you rename cart_uuid to uuid (or change uuid to cart_uuid) in the base cart module, for consistency and a nicer display in swagger?

Comment (marielejeune) [reply shopinvader_api_sale_loyalty/routers/cart.py]: Done in a new commit

**Comment by lmignon** [comment]
@marielejeune All your dev dependencies are now merged... can you cleanup the test-requirements plz?

**Review by lmignon — APPROVED** [review]

File: shopinvader_api_sale_loyalty/readme/DESCRIPTION.rst [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+Not to be confused with `shopinvader_promotion_rule`, that implements
+promotion programs from OCA module `sale_promotion_rule`, an alternative
+to the core `sale_loyalty` module.
+
+Available services:
+
+* ``/rewards/{code}`` under the ``loyalty`` router, to get all rewards claimable with a given coupon code
+* ``/apply_coupon`` under the ``cart`` router, to apply a given coupon to the cart. Allows to specify which reward and/or which free product to offer.
+* ``/apply_reward`` under the ``cart`` router, to apply a given reward (automatic promotion). Note that automatic promotions are applied automatically at cart update, when possible (if no choice must be done). This service allows to apply an automatic promotion for which the reward/free product choice is mandatory.
```
Comment (sbidoul) [inline shopinvader_api_sale_loyalty/readme/DESCRIPTION.rst]: Can you update the documentation to not reference the deprecated routes?

**Comment by marielejeune** [comment]
I've also added the `current` alias on the new cart routes, and sorted them.

**Review by lmignon — APPROVED** [review]

**Pull request merged**